### PR TITLE
Log the connection mode before disconnect and add test

### LINKED ISSUE #7391: Record connectionMode with ConnectionState_Disconnected event
As part of investigating bug https://onedrive.visualstudio.com/SPIN/_queries/edit/1184093/?triage=true, it was useful to learn how long "read" connections are. Unfortunately we do not record connectionMode property for Disconnected events as we already do not have connection when logging happens:

```typescript
    public get connectionMode(): ConnectionMode {
        if (this.connection === undefined) {
            return "read";
        }
        return this.connection.mode;
    }

    public connectionProps(): ITelemetryProperties {
        if (this.connection !== undefined) {
            return {
                sequenceNumber: this.lastSequenceNumber,
                connectionMode: this.connectionMode,
            };
        } else {
            return {
                // Report how many ops this client sent in last disconnected session
                sentOps: this.clientSequenceNumber,
            };
        }
    }
```

We can always add one more variable to track it, but I hope we could find some elegant solution to this problem without it :)

Can you please glance?

### PR DESCRIPTION
**Pull request by chensixx**
Fixes #7391 
We will now log the last connection mode right before disconnection. Test using local server to check telemetry has the right info

### CODE REVIEW

File: packages/loader/container-loader/src/deltaManager.ts [hunk 1]
```diff
@@ -397,6 +397,8 @@ export class DeltaManager
         } else {
             return {
                 ...common,
+                // last connection mode right before disconnect
+                lastConnectionMode: this.client.mode,
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:401]: Interesting. I'd just call it "connectionMode" and move it to common properties.
We record (as part of event) connectionState, so we know that it's disconnect state and I believe humans will have easy time figuring out how to interpret mode when in disconnected state.

**Review by vladsud — APPROVED** [review]

File: packages/loader/container-loader/src/deltaManager.ts (hunk 1 above)
Comment (agarwal-navin) [reply packages/loader/container-loader/src/deltaManager.ts:401]: Isn't `this.client.mode` the requested mode and not the mode in the connection? In other words, it can be different from `this.connection.mode` right? For example, client may ask for "write" mode but the connection established with the "server" is 'read"?
@vladsud

Comment (vladsud) [reply packages/loader/container-loader/src/deltaManager.ts:401]: There is this really weird line that makes it work, I think:

```typescript
                    this.client.mode = requestedMode;
                    connection = await docService.connectToDeltaStream(this.client);
```

But I agree that this is weird and can easily break.
Also it's requested mode, which can be "read", but storage can return "write".

So yes, I believe we need to add a variable to track it properly, and not rely on this.client.mode.

File: packages/loader/container-loader/src/deltaManager.ts [hunk 2]
```diff
@@ -252,6 +255,29 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
...
+     * Returns set of props that can be logged in telemetry that provide some insights / statistics
+     * about current or last connection (if there is no connection at the moment)
+    */
+    public connectionProps(): ITelemetryProperties {
+        const common = {
+            sequenceNumber: this.lastSequenceNumber,
+            connectionMode: this.lastConnectionMode,
+        };
+        if (this.connection !== undefined) {
+            this.lastConnectionMode = this.connectionMode;
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:268]: I'd not update it here. setupNewSuccessfulConnection() is much better place to do so.
Please note that you will have conflict with changes I just pushed yesterday. Please merge - some code moved here to ConnectionManager.ts

Comment (vladsud) [reply packages/loader/container-loader/src/deltaManager.ts:268]: BTW, I'd say that recording relayServiceAgent on disconnect is interesting too.
So maybe right approach is to have this._connectionProps that is empty initially, but is filled in setupNewSuccessfulConnection() with properties like connectionMode, relayServiceAgent, maybe more in future. And then this function (or rather ConnectionManager.connectionProps() simply adds it on every call)

File: packages/loader/container-loader/src/deltaManager.ts [hunk 3]
```diff
@@ -252,6 +255,29 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
             ...this.connectionManager.connectionProps,
         };
     }
+  /**
+     * Returns set of props that can be logged in telemetry that provide some insights / statistics
+     * about current or last connection (if there is no connection at the moment)
+    */
+    public connectionProps(): ITelemetryProperties {
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:262]: I believe you want to modify existing connectionProps getter above. Adding this function will do nothing as nobody will call it.

File: packages/loader/container-loader/src/deltaManager.ts (hunk 2 above)
Comment (vladsud) [reply packages/loader/container-loader/src/deltaManager.ts:268]: Also updating it here means the value for first connection will be wrong. It will record what you captured in constructor, which is not guaranteed to be what service actually returned to us.

File: packages/loader/container-loader/src/deltaManager.ts [hunk 4]
```diff
@@ -302,6 +328,7 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
             readonlyChangeHandler: (readonly?: boolean) => safeRaiseEvent(this, this.logger, "readonly", readonly),
         };
 
+        this.lastConnectionMode = this.client.mode;
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:331]: I'd rather make this.lastConnectionMode optional and have undefined as default value? This value has only indirect relationship to actual mode we will get on first connection, and (due to bug in the code or maybe by design) we can log it before connection happens. It's much more correct to see undefined than some value that is not related to actual connection modes.

Comment (chensixx) [reply packages/loader/container-loader/src/deltaManager.ts:331]: if we take the _connectionProps approach, maybe I don't need lastConnectionMode as a separate variable?

File: packages/loader/container-loader/src/connectionManager.ts [hunk 5]
```diff
@@ -646,9 +648,13 @@ export class ConnectionManager implements IConnectionManager {
             clientId: connection.clientId,
             mode: connection.mode,
         };
+
         if (connection.relayServiceAgent !== undefined) {
             this._connectionVerboseProps.relayServiceAgent = connection.relayServiceAgent;
+            this._connectionProps.relayServiceAgent = connection.relayServiceAgent;
         }
+        this._connectionProps.connectionMode = connection.mode;
```
Comment (vladsud) [inline packages/loader/container-loader/src/connectionManager.ts:656]: I'd do two minor additions:
- reset _connectionProps on line 651 to be empty, just in case
- add socketDocumentId here to this._connectionProps as well - it's in line with the how we log other properties (i.e. we will log it on disconnect as well, and it will represent state on disconnecting connection).

**Review by vladsud — APPROVED** [review]
